### PR TITLE
Adding Windows support for Travis

### LINKED ISSUE #80: Add Windows to Travis Build
Travis should test that we can make a native package on Windows
https://docs.gluonhq.com/client/0.1.33/ contains the instructions

### PR DESCRIPTION
**Pull request by mohitsaxenaknoldus**
Closes #80 

Initial commit to see how Travis responds.

### CODE REVIEW

**Comment by ghost** [comment]
### Changes to your CodeSee Architecture Map:
[View a CodeSee Map of these changes](https://app.codesee.io/r/reviews?pr=88&src=https%3A%2F%2Fgithub.com%2Ffeldoh%2FJZookeeperEdit)

Review [in an interactive map](https://app.codesee.io/r/reviews?pr=88&src=https%3A%2F%2Fgithub.com%2Ffeldoh%2FJZookeeperEdit)

View [more CodeSee Maps](https://app.codesee.io/r/maps?src=https%3A%2F%2Fgithub.com%2Ffeldoh%2FJZookeeperEdit)

#### Legend
CodeSee Map Legend

**Comment by mohitsaxenaknoldus** [comment]
Looks like Java is not supported on the Windows platform @feldoh

**Comment by feldoh** [comment]
> Looks like Java is not supported on the Windows platform @feldoh

I see you've worked it out. Yeah it's not there by default but can be choco installed. You should also be able to use graal instead of normal java but that's probably more hassle than it's worth.

I'd advise not using jdk8 though unless you have to as it's end of life. Use at least 11 so it's the same as the other version used or all the way to 17 should still work fine. 17 is latest LTS so I'd rather go for that one. Fully upgrading to 17 is a separate ticket though.

You also definitely don't want `- JAVA_HOME: /c/jdk` in the env. That won't make sense in all the environments. In fact I don't think it's valid in any of them but it's the version of java maven will use.

File: .travis.yml [hunk 1]
```diff
@@ -6,23 +6,44 @@ branches:
...
+      sudo apt install maven;
+    fi;
+    if [ "$TRAVIS_OS_NAME" = "osx" ]; then 
+      brew install maven;
+    fi;
+
 install:
   - |
-    if [[ "$TRAVIS_OS_NAME" == "osx" ]];   then DOWNLOAD_OS_NAME="darwin"; fi
-    if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then DOWNLOAD_OS_NAME="linux"; fi
```
Comment (feldoh) [inline .travis.yml:23]: you've lost the DOWNLOAD_OS_NAME so this is going to download nothing

File: .travis.yml [hunk 2]
```diff
@@ -6,23 +6,44 @@ branches:
...
 jdk:
   - openjdk11
 
 env:
   - GRAALVM_VERSION="20.2.0"
 
+before_install:
+  - |
+    if [ "$TRAVIS_OS_NAME" = "windows" ]; then 
+      choco install jdk8;
```
Comment (feldoh) [inline .travis.yml]: You don't want jdk8 here, It needs to be jdk11. This is important because the `<release>` option in the pom is set to Java 11 so you must build it with >= Java 11. Java 8 will not work. If that doesn't work on Windows (it may not set JAVA_HOME). Maven can actually run with GRAAL anyway so if you set JAVA_HOME the same as the GRAAL_HOME we set below it should build.

I have verified using GRAAL works at least on linux by using the strategy from [https://stackoverflow.com/a/49019950](https://stackoverflow.com/a/49019950) to allow testing locally without using up Travis credits. (I've also tested I can build the project from gitbash on Windows using GRAAL too just by setting JAVA_HOME to my GRAAL_HOME) We hopefully have jdk11 thanks to that openjdk11 setting above but I'm not certain what path Travis installs it to so just using GRAAL for both is probably safest.

File: .travis.yml [hunk 3]
```diff
@@ -6,23 +6,44 @@ branches:
...
-    curl -LJ "https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-$GRAALVM_VERSION/graalvm-ce-java11-$DOWNLOAD_OS_NAME-amd64-$GRAALVM_VERSION.tar.gz" --output graalvm.tar.gz
-    tar -vxzf graalvm.tar.gz
+    if [[ "$TRAVIS_OS_NAME" == "osx" ]] || [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
+        curl -LJ "https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-$GRAALVM_VERSION/graalvm-ce-java11-$DOWNLOAD_OS_NAME-amd64-$GRAALVM_VERSION.tar.gz" --output graalvm.tar.gz;
+        tar -vxzf graalvm.tar.gz;
+    fi;
+    if [[ "$TRAVIS_OS_NAME" == "windows" ]]; then
+        DOWNLOAD_OS_NAME="windows";
+        curl -LJ "https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-$GRAALVM_VERSION/graalvm-ce-java11-$DOWNLOAD_OS_NAME-amd64-$GRAALVM_VERSION.zip" --output graalvm.zip;
+        sudo apt-get install unzip -y;
```
Comment (feldoh) [inline .travis.yml]: this is windows so you want choco rather than apt and you should move it up to the install block where you're already installing maven and the jdk (though I think you can not do the jdk). I also think you might not need this at all. Because according to https://docs.travis-ci.com/user/reference/windows it comes with 7z installed. So you can do `7z.exe x graalvm.zip` this is the same as the tar command from the linux/mac versions.

**Comment by feldoh** [comment]
For reference this is the set of commands I used to debug your issue using the [docker strategy](https://stackoverflow.com/a/49019950)

You need to make sure you're setting DOWNLOAD_OS_NAME and I think the easiest option is just to set the JAVA_HOME to also use GRAAL.

```bash
git clone --depth=50 https://github.com/feldoh/JZookeeperEdit.git feldoh/JZookeeperEdit
cd feldoh/JZookeeperEdit/
git fetch origin +refs/pull/88/merge:
git checkout -qf FETCH_HEAD
export GRAALVM_VERSION="20.2.0"
export GRAALVM_HOME="$(pwd)/graalvm-ce-java11-$GRAALVM_VERSION"
export JAVA_HOME="$(pwd)/graalvm-ce-java11-$GRAALVM_VERSION"
sudo apt install maven
DOWNLOAD_OS_NAME=linux
curl -LJ "https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-$GRAALVM_VERSION/graalvm-ce-java11-$DOWNLOAD_OS_NAME-amd64-$GRAALVM_VERSION.tar.gz" --output graalvm.tar.gz;
mvn clean verify
```

Unfortunately you've used up all of the credit I have on Travis. I've put in a request for more credits but in the mean time you can test your scripts on at least Linux using that stack overflow strategy. You can also get free Windows machines on Azure if you need a local Windows environment but they are fairly terrible. Alternatively you can pr it into your own fork and have your fork built on your own Travis. You should get 10K credits for free.

On Windows in git bash

```bash
git clone --depth=50 https://github.com/feldoh/JZookeeperEdit.git feldoh/JZookeeperEdit
cd feldoh/JZookeeperEdit/
git fetch origin +refs/pull/88/merge:
git checkout -qf FETCH_HEAD
export GRAALVM_VERSION="20.2.0"
export GRAALVM_HOME="$(pwd)/graalvm-ce-java11-$GRAALVM_VERSION"
export JAVA_HOME="$(pwd)/graalvm-ce-java11-$GRAALVM_VERSION"
choco install maven
DOWNLOAD_OS_NAME=windows
curl -LJ "https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-$GRAALVM_VERSION/graalvm-ce-java11-$DOWNLOAD_OS_NAME-amd64-$GRAALVM_VERSION.zip" --output graalvm.zip;
mvn clean verify
```

File: .travis.yml (hunk 2 above)
Comment (mohitsaxenaknoldus) [reply .travis.yml]: Does that mean we don't have to install `jdk` and can use `GRAAL` if we set `GRAAL_HOME` to `JAVA_HOME`?

Comment (feldoh) [reply .travis.yml]: Almost. We don't have to install jdk if we set JAVA_HOME to GRAAL_HOME. So Atm on line 54 we have

```bash
  - export GRAALVM_HOME="$(pwd)/graalvm-ce-java11-$GRAALVM_VERSION" && if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then export GRAALVM_HOME="$GRAALVM_HOME/Contents/Home"; fi
```

You could add

```bash
  - export JAVA_HOME="$GRAALVM_HOME"
```

as line 55. And then we wouldn't need to install any JDK

File: .travis.yml [hunk 4]
```diff
@@ -31,4 +57,4 @@ script:
 
 after_success:
   - mvn clean test jacoco:report coveralls:report
-  - mvn site
+  - mvn site
```
Comment (feldoh) [inline .travis.yml]: this is complaining because it wants a blank line at the end of the file. Its a git thing

File: .travis.yml [hunk 5]
```diff
@@ -6,23 +6,49 @@ branches:
...
 
 env:
   - GRAALVM_VERSION="20.2.0"
 
+before_install:
+  - |
+    if [ "$TRAVIS_OS_NAME" = "windows" ]; then 
+      choco install jdk11;
+      choco install maven;
+      export PATH=$PATH:"/c/Program Files/Java/jdk1.8.0_191/bin";
```
Comment (feldoh) [inline .travis.yml]: this is definitely not the right path. You've installed JDK11 then added JDK8 to the path. I don't think you should actually need to add java to the path at all.

**Comment by feldoh** [comment]
@mohitsaxenaknoldus Travis has got more credit now :)

**Comment by mohitsaxenaknoldus** [comment]
> @mohitsaxenaknoldus Travis has got more credit now :)

Thanks! We are now stuck at windows not being able to find `mvn`. I've also refreshed the shell but still no help. Linux and OSX are working now.

**Comment by feldoh** [comment]
> > @mohitsaxenaknoldus Travis has got more credit now :)
> 
> Thanks! We are not stuck at windows not being able to find `mvn`. I've also refreshed the shell but still no help. Linux and OSX are working now.

Windows path stuff is a bit annoying. I suspect the issue is that because this is running in git bash refresh env won't help because it's not actually going to refresh the mingw env only the windows one.

So lets work around it. We know this is running in git bash and we know the main directory choco uses (assuming they've not changed it) so we have two options, hard code the version then hardcode the path or the better option imo is to use core bash utilities to find it and call it directly.

```bash
if [[ "$TRAVIS_OS_NAME" == "windows" ]]; then
  maven="$(find 'C:\ProgramData\chocolatey\lib\maven' -name mvn | head -n1)";
else
  maven="mvn";
fi
$maven clean verify;
...
```

Note that because we're using a variable $maven you'll have to swap all calls to `mvn` to instead be `$maven`
Also maybe add `$maven -v` so we can see which maven it's using in the output.

File: .travis.yml [hunk 6]
```diff
@@ -55,8 +55,14 @@ install:
...
+    if [[ "$TRAVIS_OS_NAME" == "windows" ]]; then
+        maven="$(find 'C:\ProgramData\chocolatey\lib\maven' -name mvn | head -n1)";
+    else
+        maven="mvn";
+    fi;
+  - $maven clean verify
+  - $maven package
 
 after_success:
   - mvn clean test jacoco:report coveralls:report
```
Comment (feldoh) [inline .travis.yml]: Also needs to be `$maven`... and you might need to do the if, find, blah in the after_success. Not sure if the after success will have retained that $maven setting. Though I think if you did `export` like the java and graal home env vars that might well carry over. You can see on the build the actual build passes but the last step, this one still triggers `mvn: command not found`

**Comment by feldoh** [comment]
Just wanted to say, well done on this! There's one remaining issue I've commented on but if you make that quick change I'm happy to approve this. It seems to work and hopefully you've enjoyed the challenge. Sorry it ended up taking so long. 41 commits (42 with that last tweak) is quite the effort! Also make sure you hit the update branch button. I'd do it myself but don't want to risk putting a commit with my name on it in your PR, just in case it interferes with your Hacktoberfest credit.

**Comment by mohitsaxenaknoldus** [comment]
> Just wanted to say, well done on this! There's one remaining issue I've commented on but if you make that quick change I'm happy to approve this. It seems to work and hopefully you've enjoyed the challenge. Sorry it ended up taking so long. 41 commits (42 with that last tweak) is quite the effort! Also make sure you hit the update branch button. I'd do it myself but don't want to risk putting a commit with my name on it in your PR, just in case it interferes with your Hacktoberfest credit.

Thank you so much! You have been a great help, pointing out many important things which I missed. Yes, this was a great challenge for me as this was my first time working with Travis. I've worked with other CICD tools but I noticed there were some different concepts in Travis. 

Also, branch updated.

**Review by feldoh — APPROVED** [review]

**Comment by feldoh** [comment]
let me know if that doesn't count
